---
title: "Example usage in R"
author: "Rutger Vos (@rvosa)"
date: "31-10-2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This document gives a first introduction in querying the TGD database such that occurrences and phylogenetic data are properly linked despite
input data potentially using different taxon labels and different backbone topologies. The overarching design of the schema is such that there
is a table of `taxa`, which uses the names accepted by either the [Mammal Species of the World 3rd edition](http://www.departments.bucknell.edu/biology/resources/msw3/), the [GBIF backbone taxonomy](https://doi.org/10.15468/39omei), or both.
As such, each record here has at least one of the identifiers used by these resources. Subsequently, all variations on these accepted names
(that is, alternative spellings, synonyms, etc.) are records in the `taxonvariants` table, each of which have a foreign key that links it to the
accepted name version. Whenever a data set (a tree, a set of occurrences, a trait data set) is imported, the record labels that are used in that
data set are looked up in the taxon variants table. Any existing exact match is re-used, and new ones are created as needed. Then, an attempt is
made to link these labels to accepted taxon names using taxonomic name reconciliation (as implemented in the `taxize` package). The upshot of this
design is that, if you want to get all occurrences (say) for a given species, you first have to look up that species in the taxa table, then get
all the variations on that name in the taxonvariants table, and then all occurrences that link to any of these variants. And, if you want to 
combine, for example, occurrences and phylogeny, you have to look for the intersection between "taxa that have variants that have occurrences",
and "taxa that have variants that have branches for the given tree".

Without further ado, here's what the steps might be. First we connect to the database:

```{r connect}
library(DBI)
tgd <- dbConnect(RSQLite::SQLite(), "../data/sql/tgd.db")
```

Then we need to look up the ID of the mammal supertree of Bininda-Emonds et al. (2007), which is named 
[../data/phylogeny/Bininda-emonds_2007_mammals.tsv](../data/phylogeny/Bininda-emonds_2007_mammals.tsv). We query as follows:

```{r tree_id_query}
tree_id <- dbGetQuery(tgd, 'select tree_id from trees where tree_name="../data/phylogeny/Bininda-emonds_2007_mammals.tsv"')$tree_id
```

Given this `tree_id`, we are now going to look for taxa that have taxonvariants that have branches in the tree and that have taxonvariants
that have occurrences. This is a fairly complex query all in all because there's a number of joins that need to be done. As follows:

```{r taxa_join}
query <- 'select distinct taxa.taxon_name,occurrences.decimal_latitude,occurrences.decimal_longitude,branches.branch_length
from (((taxonvariants 
inner join taxa on taxonvariants.taxon_id = taxa.taxon_id)
inner join occurrences on occurrences.taxonvariant_id = taxonvariants.taxonvariant_id)
inner join branches on branches.taxonvariant_id = taxonvariants.taxonvariant_id)
where branches.tree_id=12 and
taxa.taxon_level="SPECIES"
order by taxon_name'
results <- dbGetQuery(tgd, query)
```

## Statistical analysis

Now that we have a great big table, we should be able to do some potentially interesting analyses. For example:

1. Since the tropics are more biodiverse, does that mean that the species are also younger? If that's the case, then the branch lengths
   should be positively correlated with the *absolute average latitude* (i.e. distance from the equator). Of course this might also not be
   the case at all, because the underlying assumption here is that the higher biodiversity is produced by speciation rates in the tropics 
   being higher. Alternatively, extinction rates might be lower, for example.
2. Do older species have larger ranges? A very naive model of speciation and geographic distribution might be that a speciation 'event' 
   happens in a single location, and that the members of the species over time occupy a larger and larger range. If that's true (it
   probably isn't, for a variety of reasons), then one way to test this would be to calculate the *latitudinal and longitudinal range*
   (max - min, but not absolute), which again should be positively correlated with branch lengths.
3. Does [Rapoport's rule](https://en.wikipedia.org/wiki/Rapoport%27s_rule) apply to our data? This is an ecogeographical rule that states 
   that latitudinal ranges of plants and animals are generally smaller at lower latitudes than at higher latitudes. So, for this we need
   the two summary statistics from the previous two questions, i.e. the absolute average latitude and the range.

The data structure that we are going to want thus needs to have one observation for each taxon, recording 4 variables (branch length,
absolute average latitude, latitudinal range, longitudinal range). Here now we populate a data frame to do this:

```{r make_frame}
# we will use the distinct taxon names (these are the clean, accepted names) as row names
names <- unique(results$taxon_name)

# we start with a matrix, populate this, and then make it into a data.frame
df <- matrix(ncol=4, nrow=length(names))

# use i index for the matrix
for ( i in 1:length(names) ) {
	
	# average absolute latitude
	aal <- mean(abs(results[ which(results$taxon_name == names[[i]]), "decimal_latitude" ]))
	
	# latitudinal range
	lar <- range(results[ which(results$taxon_name == names[[i]]), "decimal_latitude" ])
	lar <- lar[[2]] - lar[[1]]
	
	# longitudinal range
	lor <- range(results[ which(results$taxon_name == names[[i]]), "decimal_longitude" ])
	lor <- lor[[2]] - lor[[1]]
	
	# branch length
	bl <- results[ which(results$taxon_name == names[[i]]), "branch_length" ][[1]]
	
	# add row to matrix
	df[i,] <- c( aal, lar, lor, bl )
}

# make data from, assign taxa as row names, variables as col names
df <- data.frame(df, row.names = names)
colnames(df) <- c("average_absolute_latitude", "latitudinal_range", "longitudinal_range", "branch_length")

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
